controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+        >>> rstudio = RStudio()
+        >>> rstudio.deploy_for(alice)
+        """
+
+        username = user.username.lower()
+        params = self.deploy_params(user)
+        if params:
+            params = [
+                "--set",
+                ",".join(f"{key}={val}" for key, val in params.items()),
+            ]
+        return self.helm.upgrade_release(
+            self.release_name(username),
+            self.chart_name,
+            f"--namespace={user.k8s_namespace}",
+            *params,
+        )
+
+    def get_user_deployment(self, user):
+        deployments = client.AppsV1Api().list_namespaced_deployment(
+            namespace=user.k8s_namespace,
+            label_selector=f"app={self.name}"
+        )
+
+        if len(deployments.items) < 1:
+            raise ToolNotDeployed()
+
+        return ToolDeployment(deployments.items[0], user)
+
+
+class RStudio(Tool):
+    name = "rstudio"
+
+
+class JupyterLab(Tool):
+    name = "jupyter-lab"
+
+    def release_name(self, username):
+        return f"{self.name}-{username}"
+
+    def deploy_params(self, user):
+        conf = settings.TOOLS[self.name]
+        return {
+            "Username": user.username.lower(),
+            "aws.iamRole": user.iam_role_name,
+            "toolsDomain": settings.TOOLS_DOMAIN,
+            "cookie_secret": secrets.token_hex(32),
+            "authProxy.auth0_domain": conf["domain"],
+            "authProxy.auth0_client_id": conf["client_id"],
+            "authProxy.auth0_client_secret": conf["client_secret"],
+        }
+
+
+class ToolDeployment(object):
+
+    def __init__(self, deployment, user):
+        self.deployment = deployment
+        self.user = user
+
+    @classmethod
+    def list(cls, user):
+        deployments = client.AppsV1Api().list_namespaced_deployment(
+            user.k8s_namespace,
+        )
+        return [
+            ToolDeployment(deployment, user) for deployment in deployments.items
+        ]
+
+    @property
+    def name(self):
+        return self.deployment.metadata.labels["app"]
+
+    @property
+    def username(self):
+        return self.user.username.lower()
+
+    @property
+    def url(self):
+        return f"https://{self.username}-{self.name}.{settings.TOOLS_DOMAIN}"
+
+    @property
+    def pods(self):
+        pods = client.CoreV1Api().list_namespaced_pod(
+            self.user.k8s_namespace,
+            label_selector=f"app={self.name}",
+        )
+        return [Pod(pod) for pod in pods.items]
+
+    @property
+    def available(self):
+        return any(pod.status["phase"] == "Running" for pod in self.pods)
+
+    @property
+    def idled(self):
+        return self.deployment.metadata.labels.get(IDLED) == "true"
+
+    def restart(self):
+        return client.AppsV1Api().delete_collection_namespaced_replica_set(
+            namespace=self.user.k8s_namespaces,
+            label_selector=f"app={self.name}",
+            watch=True,
+        )
```

Comment:
#### 1e. Refactor Tools code
>Lines 158-206

The `ToolDeployment` class is a wrapper around the Kubernetes Deployment for a specified tool.

It provides a class method to get a list of all a user's tool Deployments, a property to list a specific Deployment's Pods, and a method to restart a tool (by deleting all the pods, prompting Kubernetes to spin up new ones).

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-aad7b41260663fb400ff942a02fd3d14R209)